SplitWebKeys

jtmcdole opened the pull request:
Breaks out changes from #3 and does an overhaul:
 - Break apart public/private key pairs
 - Split up the library for easier groking.
 - Adds the example from rfc8291 section 5 (sanity check)

(will close out #3 to try and keep things clean)

jtmcdole committed:
Add WebPush encryption support

jtmcdole committed:
WebPushKeys can be "public"

Servers will only get the public+auth from the client.

jtmcdole committed:
Upgrade VM and unpin webcrypto

Breakout of https://github.com/UnifiedPush/dart-webpush-encryption/pull/3

jtmcdole committed:
Merge remote-tracking branch 'origin/main' into splitWebKeys

jtmcdole committed:
Add encryption support (part 2)

Breaks out changes from #3 and does an overhaul:
 - Break apart public/private key pairs
 - Split up the library for easier groking.
 - Adds the example from rfc8291 section 5 (sanity check)

jtmcdole committed:
Update pub version

p1gp1g commented on lib/src/keys.dart:
```diff
@@ -0,0 +1,212 @@
+/// Stores the public, private, auth key needed for webpush.
+class WebPushKeys {
+  final WebPublicPushKey? publicKey;
+  final WebPrivatePushKey? privateKey;
```
```suggestion
  final PublicWebPushKey publicKey;
  final PrivateWebPushKey privateKey;
```

I suggest to:
- Rename the variables so "WebPush" is not splitted. Public and Private are set in the beginning to make it clearer.
- Force having both. A key set is a pubkey + privkey

p1gp1g commented on lib/webpush_encryption.dart:
```diff
@@ -42,106 +43,49 @@ class WebPush {
+  /// if it is not provided. This is mostly useful for debugging.
+  static Future<Uint8List> encrypt({
+    required WebPushKeys serverKeys,
+    required WebPushKeys clientKeys,
```
```suggestion
    required PublicWebPushKey clientPubKey,
```

Here we only need the pubKey

p1gp1g commented on lib/webpush_encryption.dart:
```diff
@@ -42,106 +43,49 @@ class WebPush {
+        fillRandomBytes(salt);
+      }
+      final uaPubKey = await clientKeys.publicKey!.pubKey;
+      final asPubKey = await serverKeys.publicKey!.pubKey;
```
```suggestion
      final asPubKey = await serverKeys.publicKey.pubKey;
```

As we have the server's key set, there will always be a pubkey and a privkey

p1gp1g replied:
Todo in other occurrences of course

p1gp1g reviewed (changes requested):
Thank you for the PR !

So I suggest to rename the pubkey and privkey classes `PublicWebPushKey` and `PrivateWebPushKey` to not split `WebPush` and highlight that its a private or public key

I think we should rename `WebPushKeys` to `WebPushKeyPair` or `WebPushKeySet`

And a _keyset_ should always have both _pub_ and _privkey_. When a function needs a _pubkey_, it has to request only the _pubkey_. When it needs a _privkey_, most of the time it also uses the _pubkey_, so the function requests the _keyset_.

jtmcdole committed:
Address comments

jtmcdole committed:
More updates:

- remove static methods from WebPush for future testing (you can mock
  them).
- Update readme.md with changes
- address comments

jtmcdole commented:
Updated from comments, updated README,  removed some static methods for future testing (mocking) support.

jtmcdole committed:
IMPORTANT note to >

jtmcdole committed:
Another note change

p1gp1g commented on README.md:
```diff
@@ -5,32 +5,31 @@ This is primarily an RFC 8291 (Message Encryption for Web Push) implementation,
+See [example/simple.dart](example/simple.dart).
 
-There are two main classes here, WebPushKeys, and WebPush.
+There are three main classes here, `WebPushKeySet`, `PublicWebPushKey`, `WebPush`.
```
```suggestion
There are 4 main classes here, `WebPushKeySet`, `PublicWebPushKey`, `PrivateWebPushKey`, `WebPush`.
```

p1gp1g commented on README.md:
```diff
@@ -5,32 +5,31 @@ This is primarily an RFC 8291 (Message Encryption for Web Push) implementation,
-1. Use `.newKeyPair()` generate a new key.
+In `WebPushKeySet`:
+1. Use `.newKeyPair()` generate a new key pair (private & public).
 1. Use `.p256dh` and `.auth` to export the keys into the base64 encoding that can be sent to a server.
```
This should be in (missing) `PublicWebPushKey`, the doc for `PrivateWebPushKey is also missing

p1gp1g reviewed (changes requested):
There are small changes to do on the readme. Can you explain why it is better to remove static methods ?

After that, it looks good

jtmcdole committed:
Update README.md

Co-authored-by: S1m <31284753+p1gp1g@users.noreply.github.com>

jtmcdole commented:
>  Can you explain why it is better to remove static methods?

Makes it easier for downstream users to mock WebPush using mockito or mocktail packages. If you want quick access without storing a simple/cheap class allocation, a top-level `const webPush = WebPush()` would work as well - similar to the top level `utf8` class in dart:convert

jtmcdole committed:
dartdoc for public/private keys

p1gp1g commented on README.md:
```diff
@@ -5,32 +5,31 @@ This is primarily an RFC 8291 (Message Encryption for Web Push) implementation,
 1. Use `.p256dh` and `.auth` to export the keys into the base64 encoding that can be sent to a server.
 1. Use `.serialize` to serialize the public *AND private keys* for storage. Be careful with this since it contains the private key.
    Then, use `WebPushKeys.deserialize` to deserialize that back into an object from the stored string.
 
```
```suggestion
1. Use `.newKeyPair()` generate a new key pair (private & public).
1. Use `.publicKey` to get the `PublicWebPushKey` of the key set.
1. Use `.serialize` to serialize the public *AND private keys* for storage. **Be careful with this since it contains the private key.**
   Then, use `WebPushKeys.deserialize` to deserialize that back into an object from the stored string.

In `PublicWebPushKey`:
1. Use `.p256dh` and `.auth` to export the keys into the base64 encoding that can be sent to a server.
```

p1gp1g commented:
Thank you for the explanation, I don't know if @karmanyaahm you are available for a review ? I will merge it else :)

jtmcdole committed:
Update README.md

Co-authored-by: S1m <31284753+p1gp1g@users.noreply.github.com>

karmanyaahm commented:
I took a brief look at it and it looks good to me

p1gp1g commented:
So I'm merging it. Thank you for the contribution !

I'll publish it to pubdev during the week, feel free to ping me if I forget

jtmcdole commented:
Excellent! Can't wait to see it up on pub.dev.  I guess tests are next up.
